osquery/core/posix/utils.cpp

```diff
@@ -41,4 +41,35 @@ Status platformStrncpy(char* dst, size_t nelms, const char* src, size_t count) {
   ::strncpy(dst, src, count);
   return Status(0, "OK");
 }
+
+std::string canonicalize_file_name(char* name) {
+  char* buffer = nullptr;
+#ifdef PATH_MAX
+  // On supported platforms where PATH_MAX is defined we can pass null
+  // as buffer, and allow libc to alloced space
+  // On centos/ubuntu libc will use realloc so we will be able to resolve
+  // any path
+  // On darwin libc will allocate PATH_MAX buffer for us
+  buffer = nullptr;
+  char* resolved = realpath(name, nullptr);
+  std::string result = (resolved == nullptr) ? name : resolved;
+  free(resolved);
+#else
+#warnign PATH_MAX is undefined, please read comment below
+  // PATH_MAX is not defined, very likely it's not officially supported
+  // os, our best guess is _PC_PATH_MAX if available
+  // In case of failure fallback to "safe" buffer of 8K
+
+  long int path_max = pathconf(name, _PC_PATH_MAX);
+  if (path_max <= 0) {
+    path_max = 8 * 1024;
+  }
+  char* buffer = reinterpret_cast<char*>(malloc(path_max));
+  char* resolved = realpath(name, buffer);
```

Comment:
Code looks more complex with code reusage here, the aim is to separate them, clearly idicate that we are using nullptr in first case. IMHO it will be easier to maintain this code in case if we will change fallback code to custom  implementaion.